internal/controllers/app_controller.go

```diff
@@ -270,12 +290,245 @@ func (r *AppReconciler) reconcile(ctx context.Context, app *ketchv1.App, logger
+// with updated deployment status and/or healthcheck and timeout failures
+func (r *AppReconciler) watchDeployEvents(ctx context.Context, app *ketchv1.App, namespace string, dep *appsv1.Deployment, process *ketchv1.ProcessSpec, recorder record.EventRecorder) error {
+	config, err := GetRESTConfig()
+	if err != nil {
+		return err
+	}
+	cli, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	opts := metav1.ListOptions{
+		FieldSelector: "involvedObject.kind=Pod",
+		Watch:         true,
+	}
+	watcher, err := cli.CoreV1().Events(namespace).Watch(ctx, opts) // requires "watch" permission on events in clusterrole
+	if err != nil {
+		return err
+	}
+	defer watcher.Stop()
+
+	// wait for Deployment Generation
+	timeout := time.After(DefaultPodRunningTimeout)
+	for dep.Status.ObservedGeneration < dep.Generation {
+		dep, err = cli.AppsV1().Deployments(namespace).Get(ctx, dep.Name, metav1.GetOptions{})
+		if err != nil {
+			recorder.Eventf(app, v1.EventTypeWarning, appReconcileError, "error getting deployments: %s", err.Error())
+			return err
+		}
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-timeout:
+			recorder.Event(app, v1.EventTypeWarning, appReconcileError, "timeout waiting for deployment generation to update")
+			return errors.Errorf("timeout waiting for deployment generation to update")
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	return r.watchFunc(ctx, app, namespace, dep, process, recorder, watcher, cli, timeout)
+}
+
+func (r *AppReconciler) watchFunc(ctx context.Context, app *ketchv1.App, namespace string, dep *appsv1.Deployment, process *ketchv1.ProcessSpec, recorder record.EventRecorder, watcher watch.Interface, cli kubernetes.Interface, timeout <-chan time.Time) error {
+	var err error
+	watchCh := watcher.ResultChan()
+	recorder.Eventf(app, v1.EventTypeNormal, appReconcileStarted, "Updating units [%s]", process.Name)
+
+	var specReplicas int32
+	if dep.Spec.Replicas != nil {
+		specReplicas = *dep.Spec.Replicas
+	}
+	oldUpdatedReplicas := int32(-1)
+	oldReadyUnits := int32(-1)
+	oldPendingTermination := int32(-1)
+	now := time.Now()
+	var healthcheckTimeout <-chan time.Time
+
+	go func() {
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case msg, isOpen := <-watchCh:
+			if !isOpen {
+				return
+			}
+			if isDeploymentEvent(msg, dep) {
+				recorder.Eventf(app, v1.EventTypeNormal, appReconcileUpdate, "%s", stringifyEvent(msg))
+			}
+		case <-healthcheckTimeout:
+			err = createDeployTimeoutError(ctx, cli, app, time.Since(now), namespace, "healthcheck")
+			recorder.Eventf(app, v1.EventTypeWarning, appReconcileError, "error waiting for healthcheck: %s", err.Error())
+			return
+		case <-timeout:
+			err = createDeployTimeoutError(ctx, cli, app, time.Since(now), namespace, "full rollout")
+			recorder.Eventf(app, v1.EventTypeWarning, appReconcileError, "deployment timeout: %s", err.Error())
+			return
+		case <-ctx.Done():
+			return
+		}
+	}()
+
+	for {
```

Comment:
will it always break?

do we always either get and error or the following condition is always met? 
```
readyUnits == specReplicas &&
dep.Status.Replicas == specReplicas
```


should we have timeout?

Reply:
Good call. I re-added the bit that checks for timeouts.